motion_tracker/utils/image_generator.py

```diff
@@ -4,212 +4,195 @@
+class ImagePairAndBoxGen(object):
+    def __init__(self, crops_per_image = 10, batch_size=50,
+                 output_width = 256, output_height = 256,
+                 desired_dim_ordering='th'):
+        """Generator yielding dictionary of image and bounding boxes.
+        This is an abstract class and will not be called directly.
+
+        Args:
+        ----
+            crops_per_image: int
+                Number of random crops of the current frame to take.
+                Corresponds to both k3 and k4 in the paper, with a value of 10.
+            batch_size: int
+                Number of items in list to be returned.
+            output_width: int
+                Width in pixels of output images for start and ending image
+            output_height: int
+                Height in pixels of output images for start and ending image
+            desired_dim_ordering: Either 'th' or 'tf' to determine where channel index is
+        """
+        self.crops_per_image = crops_per_image
+        self.batch_size = batch_size
+        self.output_width = output_width
+        self.output_height = output_height
+        self.desired_dim_ordering = desired_dim_ordering
+
+
+    def are_accumulators_full(self):
+        return self.X['start_img'].shape[0] >= self.batch_size
+
+    def flow(self):
+        """The function yielding an iterator."""
+        while True:
+            self.reset_accumulators()
+            while not self.are_accumulators_full():
+                self.add_to_accumulators()
+            self.remove_accumulator_overflow()
+            self.ensure_dim_ordering()
+            yield self.X, self.y
+
+
+    def reset_accumulators(self):
+        """Empties all contents from previous batch. Sets up for next batch"""
+        empty_img_accumulator = np.zeros([0, self.output_width, self.output_height, 3]).astype('uint8')
+        empty_box_accumulator = np.zeros([0, 4]).astype('uint16')
+        empty_digit_accumulator = np.zeros([0]).astype('uint16')
+        self.X = {'start_img': empty_img_accumulator,
+                  'start_box': empty_box_accumulator,
+                  'end_img': empty_img_accumulator}
+        self.y = {'x0': empty_digit_accumulator,
+                  'y0': empty_digit_accumulator,
+                  'x1': empty_digit_accumulator,
+                  'y1': empty_digit_accumulator}
+
+    def remove_accumulator_overflow(self):
+        """Cuts extra data if we have more data than batch size"""
+        for field in self.X:
+            self.X[field] = self.X[field][:self.batch_size]
+        for field in self.y:
+            self.y[field] = self.y[field][:self.batch_size]
+
+    def ensure_dim_ordering(self):
+        if self.desired_dim_ordering == 'th': # swap because data comes in 'tf' format
+            self.X['start_img'] = self.X['start_img'].swapaxes(1, 3)
+            self.X['end_img'] = self.X['end_img'].swapaxes(1, 3)
+
+    def normalize(self, img):
+        return (img-100) / 60
+
```

Comment:
It's also a very uncommon way to do normalization... normally people subtract off the average (over many images) of each pixel/channel value.  That potentially could matter, but is very close to what I actually did, since those values were all very close to each other.